[IN DISCUSSION] Provide packages for different NVidia driver types - mediated

jimklimov opened the pull request:
Alternative/follow-up to #3458 with proposed mediation of FS objects delivered by each driver in such a way that they can be installed all at once (or in any combination), and then the user can pick one to be enabled via IPS mediation.

NOTE: Still not quite sure how to properly address the driver_aliases in a way that drivers won't conflict over common cards, and won't claim ones they do not know either. And also having the active driver named `nvidia` is probably still a goal.

Hopefully would supersede the PR #5337 as well ;)

To build this in a local testing environment, you'd probably need to set up a local publisher, `make publish` the `nvidia-select` into it, and then piggy-back on encumbered builds support to keep the `pkglint` satisfied while publishing the actual driver packages and the `nvidia` metapackage (new package, old name). So something like:

````
:; ( cd components/openindiana/nvidia-select && make publish )
:; for N in 304 340 084 387 390 396 ; do ( \
    cd components/openindiana/nvidia-$N && make ENCUMBERED=yes publish ; \
   ) ; done
:; ( cd components/meta-packages/nvidia && make ENCUMBERED=yes publish )
````

and then you should be able to install them into a local test BE which would use this publisher (possibly using the full FMRIs to publisher+package names).

jimklimov committed:
Renamed NVidia-340.x component dir and added an NVidia-384.x component dir, and a "nvidia" metapackage to require either of these variants

jimklimov committed:
Typo fixes in copyright lines (extra whitespace)

jimklimov committed:
driver-graphics-nvidia-340.p5m : match formatting and default rights with nvidia-384 to simplify diffs

jimklimov committed:
nvidia driver packages : update the wording of CDDL headers

jimklimov committed:
Add nvidia-304 for yet older GeForce cards (6-/7-series)

Commit:
Bump minor versions of nvidia drivers

Commit:
Fix wrong pci ids of nvidia drivers 304 and 384

jimklimov committed:
Merge pull request #1 from optarg/nvidia

Some minor fixes to your NVIDIA driver package PR

alarcher committed:
Bump Nvidia to 340.104

jimklimov committed:
Merge remote-tracking branch 'upstream/oi/hipster' into nvidia-multi

jimklimov committed:
nvidia-304/Makefile : introduce NVIDIA_DRIVER_MAJOR_VERSION, NVIDIA_DRIVER_MINOR_VERSION and NVIDIA_MEDIATOR

jimklimov committed:
nvidia-304/Makefile : move preparation of PROTO_DIR with version-named files and dirs into Makefile, break up the big install ruleset into better visible steps, and symlink back to original names to facilitate manifest generation

jimklimov committed:
nvidia-304 sample-manifest.p5m regenerated with new recipe

jimklimov committed:
nvidia-304/driver-graphics-nvidia-304.p5m : introduce NVIDIA_DRIVER_MAJOR_VERSION

jimklimov committed:
nvidia-304 : revised driver-graphics-nvidia-304.p5m for mediation-readiness

jimklimov committed:
nvidia-304/Makefile : introduce sample-manifest-nvidia target and sample-manifest.p5m.nvidia generated with the new recipe

jimklimov committed:
nvidia-384 : bump to 384.98

jimklimov committed:
nvidia-384/Makefile : introduce NVIDIA_DRIVER_MAJOR_VERSION, NVIDIA_DRIVER_MINOR_VERSION and NVIDIA_MEDIATOR

jimklimov committed:
nvidia-384/Makefile : introduce sample-manifest-nvidia target and sample-manifest.p5m.nvidia generated with the new recipe

jimklimov committed:
nvidia-304/Makefile : mention alternate possible download links for posterity

jimklimov committed:
nvidia-384/Makefile : move preparation of PROTO_DIR with version-named files and dirs into Makefile, break up the big install ruleset into better visible steps, and symlink back to original names to facilitate manifest generation

jimklimov committed:
nvidia-384 sample-manifest.p5m regenerated with new recipe

jimklimov committed:
nvidia-384/driver-graphics-nvidia-384.p5m : introduce NVIDIA_DRIVER_MAJOR_VERSION

jimklimov committed:
nvidia-384 : revised driver-graphics-nvidia-384.p5m for mediation-readiness

jimklimov committed:
nvidia-340/Makefile : introduce NVIDIA_DRIVER_MAJOR_VERSION, NVIDIA_DRIVER_MINOR_VERSION and NVIDIA_MEDIATOR

jimklimov committed:
nvidia-340/Makefile : mention alternate possible download links for posterity

jimklimov committed:
nvidia-340/Makefile : introduce sample-manifest-nvidia target and sample-manifest.p5m.nvidia generated with the new recipe

jimklimov committed:
nvidia-340/Makefile : move preparation of PROTO_DIR with version-named files and dirs into Makefile, break up the big install ruleset into better visible steps, and symlink back to original names to facilitate manifest generation

jimklimov committed:
nvidia-340/driver-graphics-nvidia-340.p5m : introduce NVIDIA_DRIVER_MAJOR_VERSION

jimklimov committed:
nvidia-340 sample-manifest.p5m regenerated with new recipe

jimklimov committed:
nvidia-340 : revised driver-graphics-nvidia-340.p5m for mediation-readiness

jimklimov committed:
Add nvidia-387 (387.12)

jimklimov commented:
Added recipe for nvidia-387.12 (the newest maintained driver line), mostly cloned from recipe for nvidia-384. Probably the driver_aliases need to be updated with hardware support introduced with this line.

PRs against my branch are welcome :)

kenmays commented:
1. Replace 387.12 with 387.22 driver. This is new primary driver. GeForce 400-series and higher
2. Limit 340.104 driver PCI IDs to support only <= GeForce 300-series. This is old primary driver.
3. Limit 304.137 driver PCI IDs to support only <= GeForce 7-series. This is legacy driver.

This will allow proper fallback between drivers.
Packaging setup conventions:
driver/graphics/nvidia - 387.22
driver/graphics/nvidiaR340 - 340.104
driver/graphics/nvidiaR304 - 304.137

jimklimov committed:
Bump nvidia-387 to 387.22

jimklimov committed:
nvidia-*/Makefile : use print-pci-ids.sh to list PCI IDs into sample-manifest.p5m.nvidia

jimklimov committed:
nvidia-387/sample-manifest.p5m.nvidia : update with PCI IDs supported by this driver

jimklimov committed:
nvidia-*/print-pci-ids.sh : avoid trailing backslash on last line

jimklimov committed:
driver-graphics-nvidia-387.p5m : mediate dev/nvidia_modeset

jimklimov committed:
driver-graphics-nvidia-387.p5m : update driver_aliases

jimklimov committed:
driver-graphics-nvidia-384.p5m : mediate dev/nvidia_modeset

jimklimov committed:
nvidia-340/sample-manifest.p5m.nvidia : update with PCI IDs supported by this driver

jimklimov committed:
nvidia-304/sample-manifest.p5m.nvidia : update with PCI IDs supported by this driver

jimklimov committed:
nvidia-384/sample-manifest.p5m.nvidia : update with PCI IDs supported by this driver

jimklimov committed:
driver-graphics-nvidia-384.p5m : update driver_aliases

jimklimov commented:
The `driver_aliases` were updated with whatever HW support each driver release claims, though not so much was lagging behind. Also automated preparation of this part of the `sample-manifest.p5m.nvidia` files to facilitate future updates of the drivers.

People who know NVidia branding, as well as nuances and limitations of certain models and drivers, are welcome to refine the list ;) See also http://pciids.sourceforge.net/v2.2/pci.ids (thanks to @kenmays)

jimklimov committed:
driver-graphics-nvidia-*.p5m : (hard)link formatting, more care about mediation

jimklimov committed:
nvidia* : "legacy" can not be mediated, move it to metapackage

jimklimov commented:
@kenmays : Refined this PR to the point that all 5 (4 drv + 1 meta) packages can almost be co-installed; one issue remains : 
````
pkg install: The following packages all deliver driver actions to nvidia: (...)
````
-- so your suggestion about keeping the PCI IDs non-overlapping is correct... Perhaps, addition-removal of driver_aliases entries can be put into some service to which a file will be fed as part of mediation switcheroo?

Larger install log for context:
````
root@jimoi:/root# pkg -R /a refresh oi-userland-jim

root@jimoi:/root# pkg -R /a install --no-refresh -v pkg://oi-userland-jim/driver/graphics/nvidia{,-340,-304,-384,-387}

Creating Plan (Checking for conflicting actions): /
  Packages to install:  4
  Packages to update:  1
Rebuild boot archive: No

Changed packages:
oi-userland-jim
  driver/graphics/nvidia-304
    None -> 0.304.137-2017.0.0.0
  driver/graphics/nvidia-340
    None -> 0.340.104-2017.0.0.0
  driver/graphics/nvidia-384
    None -> 0.384.98-2017.0.0.0
  driver/graphics/nvidia-387
    None -> 0.387.22-2017.0.0.0
openindiana.org -> oi-userland-jim
  driver/graphics/nvidia
    0.340.104-2017.0.0.0 -> 1.0.0-2017.0.0.0

pkg install: The following packages all deliver driver actions to nvidia:

  pkg://oi-userland-jim/driver/graphics/nvidia-340@0.340.104,5.11-2017.0.0.0:20171113T005449Z
  pkg://oi-userland-jim/driver/graphics/nvidia-304@0.304.137,5.11-2017.0.0.0:20171113T005458Z
  pkg://oi-userland-jim/driver/graphics/nvidia-387@0.387.22,5.11-2017.0.0.0:20171113T005328Z
  pkg://oi-userland-jim/driver/graphics/nvidia-384@0.384.98,5.11-2017.0.0.0:20171113T005440Z

These packages may not be installed together. Any non-conflicting set may
be, or the packages must be corrected before they can be installed.
````

jimklimov committed:
Add nvidia-select package

jimklimov committed:
nvidia-*/Makefile : simplify /usr/**/nvidia and /usr/**/NVIDIA move+symlinking

jimklimov committed:
nvidia-*/p5m : remove conflicting driver actions

jimklimov committed:
nvidia-*/p5m : depend on pkg:/driver/graphics/nvidia-select as a means of installing driver aliases

jimklimov committed:
nvidia-select : add a service to refresh or ensure the selected NVidia driver version

jimklimov commented:
It installs!

````
root@jimoi:/root# pkg -R /a install --no-refresh -v pkg://oi-userland-jim/driver/graphics/nvidia{,-340,-304,-384,-387}
           Packages to install:        5
            Packages to update:        1
           Mediators to change:        1
            Services to change:        4
     Estimated space available: 26.09 GB
Estimated space to be consumed:  1.55 GB
          Rebuild boot archive:      Yes

Changed mediators:
  mediator nvidia-drv:
           version: None -> 387 (system default)
    implementation: None -> nvidia-drv-proprietary (system default)

Changed packages:
oi-userland-jim
  driver/graphics/nvidia-304
    None -> 0.304.137-2017.0.0.0
  driver/graphics/nvidia-340
    None -> 0.340.104-2017.0.0.0
  driver/graphics/nvidia-384
    None -> 0.384.98-2017.0.0.0
  driver/graphics/nvidia-387
    None -> 0.387.22-2017.0.0.0
  driver/graphics/nvidia-select
    None -> 1.0.0-2017.0.0.0
openindiana.org -> oi-userland-jim
  driver/graphics/nvidia
    0.340.104-2017.0.0.0 -> 1.0.0-2017.0.0.0

Services:
  refresh_fmri:
    svc:/system/nvidia-select
  restart_fmri:
    svc:/application/desktop-cache/desktop-mime-cache:default
    svc:/application/desktop-cache/icon-cache:default
    svc:/system/manifest-import:default

Editable files to change:
  Install:
    kernel/drv/nvidia-304.conf
    kernel/drv/nvidia-340.conf
    kernel/drv/nvidia-384.conf
    kernel/drv/nvidia-387.conf
    kernel/drv/nvidia_modeset-384.conf
    kernel/drv/nvidia_modeset-387.conf

Planning linked: 0/5 done; 1 working: zone:selftest-giz-02n
Linked image 'zone:selftest-giz-02n' output:
|      Estimated space available:  26.09 GB
| Estimated space to be consumed: 308.48 MB
|           Rebuild boot archive:        No
`
(...)
Planning linked: 5/5 done
DOWNLOAD                                PKGS         FILES    XFER (MB)   SPEED
Completed                                6/6       422/422  240.7/240.7    0B/s

Downloading linked: 0/5 done; 1 working: zone:(...)
Downloading linked: 5/5 done
PHASE                                          ITEMS
Removing old actions                         146/146
Installing new actions                       678/678
Updating modified actions                      11/11
Updating package state database                 Done 
Updating package cache                           1/1 
Updating image state                            Done 
Creating fast lookup database                working 
````

Whether it works is a different matter, though now quite open to experimentation ;)

kenmays commented:
Ref: https://www.freebsd.org/cgi/ports.cgi?query=nvidia&stype=name
Looks great otherwise.

jimklimov commented:
> Why not just use rem_drv / add_drv?

This logic is more or less a clone of one provided in the SVR4 driver which is assumed to work :)
Also, I'm not sure `add_drv` will properly add all the pci-id based aliases semi-randomly peppered over these drivers ;) So far I did not actually run much of this, and the night was late and dark, so I could have screwed stuff up.

> Do we create /reconfigure each time we run this script?

The original postinstall/postremove set of scripts does. Here the script started out as something to run manually (so yes, the user wants to reconfigure) and progressed into a handler for SMF service too - so in this case essentially there's nothing to do if start'ing with the same driver as the last time - installed and chosen, and there's reconfiguration due for auto-detection if the driver was not configured; my understanding is that addition/removal/upgrade of the nvidia-XYZ packages will tickle the refresh action to re-detect a best fit driver and remove/re-add the known supported PCI IDs.

I see many ways the logic in the script can be improved; so far it was primarily a way to become able to install the four packages at once and yet pick one driver to use along with those PCI IDs it supports per original SVR4 packaging metadata.

> Do we miss a service to run it on boot?

There is a `nvidia-select` service registered.... unless I missed something (I could, it was very late in the night) it should run quite on boot - before X11 and before ogl-select (to mediate the driver and libraries version, if needed) explicitly in particular.

jimklimov committed:
nvidia-select : allow to --update-smf with interactive usage too

jimklimov committed:
nvidia-select : track the requested USER_ACTION to not just rely on value of NVIDIA_VERSION

jimklimov committed:
nvidia-select : refresh after removal of selected driver should not die() but try to pick another

jimklimov committed:
nvidia-select : a bit more verbosity to track what happens in service or CLI invokation logs

jimklimov committed:
nvidia-select : bump service timeouts

jimklimov committed:
nvidia-select : check if mediator is already set to what we want - this is quicker and safer (set-mediator returns an error exit code otherwise)

jimklimov committed:
nvidia-select : fix svccfg addpg calls

jimklimov committed:
nvidia-select : strip markup cruft from aliases list

jimklimov committed:
nvidia-select : try to reload the kernel driver too

jimklimov committed:
nvidia-select : in auto mode, select old default and only version: nvidia-340

jimklimov committed:
nvidia-select : rearrange set-mediator command

jimklimov committed:
nvidia-select : rearrange reloading a kernel module, and try to reload all involved (nvidia, nvidia_modeset, gfx_private)

jimklimov committed:
nvidia-XXX : add a symlink for /usr/lib/xorg/modules/drivers/nvidia_drv.so - it happens to be less confusing for our driver

jimklimov committed:
nvidia-select : finish with a few suggestions

jimklimov committed:
nvidia-select : suggest the current value of ogl-select vendor in the end

jimklimov committed:
nvidia-select.xml : typo fix

jimklimov commented:
Revised the code for this PR by actually booting into a BE with this installed, found a few improvements and even fixes here and there. Alas, on my particular notebook `NVidia_drv` fails to load in XOrg probably due to Optimus coexistence with an intel chip, so I can't actually test how drivers work. 

But mediated installation of the bunch and mediated selection of a desired driver version does work.

alarcher committed:
nvidia: bump to 340.106

alarcher committed:
nvidia: bump to 304.107

Mno-hime committed:
10913 ELF file /a/kernel/drv/nvidia is 32-bit and will be excluded
11075 OpenGL headers get removed after removing driver/graphics/nvidia for update

jimklimov committed:
Temporarily `git mv nvidia-340 nvidia` to resync with upstream/oi/hipster

jimklimov committed:
Merge commit '2136604c82fa97645ec6488d4a395fdaebb7eba4' into nvidia-multi-mediated

jimklimov committed:
Merge commit '93c201309285e723e6a141c48ec4d15098866289' into nvidia-multi-mediated

jimklimov committed:
Merge commit '96984659a921be0d38c0212b02eabaa84808206a' into nvidia-multi-mediated

jimklimov committed:
nvidia-304/384/387 : Convert gcc-4-runtime to gcc-6-runtime

jimklimov committed:
nvidia-304/384/387 : use GCC 6 runtime libraries

jimklimov committed:
nvidia-304/340/384/387 : use GCC 6 runtime libraries in a driver-version pathname dependent manner

jimklimov committed:
nvidia-304/384/387 : Use GCC_RUNTIME_PKG in REQUIRED_PACKAGES

jimklimov committed:
Temporarily `git mv nvidia/*-340* nvidia/*` to resync with upstream/oi/hipster

jimklimov committed:
Rename back "nvidia" to "nvidia-340"

jimklimov committed:
nvidia-304 : reapply changes from 10913 ELF file /a/kernel/drv/nvidia is 32-bit and will be excluded and 11075 OpenGL headers get removed after removing driver/graphics/nvidia for update

jimklimov committed:
nvidia-340 : reapply symlink-related changes from 10913 ELF file /a/kernel/drv/nvidia is 32-bit and will be excluded and 11075 OpenGL headers get removed after removing driver/graphics/nvidia for update

jimklimov committed:
nvidia-304 : reapply symlink-related changes from 10913 ELF file /a/kernel/drv/nvidia is 32-bit and will be excluded

jimklimov committed:
nvidia-340 : reapply symlink-related changes from 10913 ELF file /a/kernel/drv/nvidia is 32-bit and will be excluded

jimklimov committed:
nvidia-384 : reapply changes from 10913 ELF file /a/kernel/drv/nvidia is 32-bit and will be excluded and 11075 OpenGL headers get removed after removing driver/graphics/nvidia for update

jimklimov committed:
nvidia-387 : reapply changes from 10913 ELF file /a/kernel/drv/nvidia is 32-bit and will be excluded and 11075 OpenGL headers get removed after removing driver/graphics/nvidia for update

jimklimov committed:
Temporarily move nvidia-340 to nvidia to resync with upstream branch again

jimklimov committed:
Merge remote-tracking branch 'upstream/oi/hipster' into nvidia-multi-mediated

jimklimov committed:
Rename back "nvidia" to "nvidia-340" again

jimklimov commented:
Following up on recent discussions in the mailing list, this PR has now been updated vs. changes that happened in `upstream/oi/hipster` over the past couple of years.

jimklimov commented:
According to https://www.nvidia.com/en-us/drivers/unix/solaris-display-archive/ the currently newest versions of drivers in this bunch are:

* nvidia-304 - 304.137 from September 19, 2017
* nvidia-340 - 340.107 from June 6, 2018
* nvidia-384 - 384.130 from March 28, 2018
* nvidia-387 - 387.34 from November 24, 2017

There are also a number of newer (and older FWIW) series that can be added easily into this design by cloning the recipes and adapting for file lists, e.g. kernel module sets and names, and editing components/meta-packages/nvidia/driver-graphics-nvidia.p5m listing.

jimklimov commented:
I'll update the PR for the currently latest 384 and 387 drivers (waiting for downloads to complete so I can write the checksums), and hope to add a nvidia-390 per that e-mail discussion : https://www.mail-archive.com/openindiana-discuss@openindiana.org/msg22695.html

3eka commented:
Hi Jim, 
thanks for update  on this.

BTW: https://wiki.openindiana.org/oi/Nvidia+Graphics shows (and offers for download) 396.24 version.

Regards.

jimklimov committed:
nvidia-387 : bump to 387.34

jimklimov committed:
nvidia-384 : bump to 384.130

jimklimov committed:
nvidia-304/340/384/387 : bump sample-manifest.p5m.nvidia about 32-bit driver removals

jimklimov committed:
nvidia-304/340/384/387 : bump Makefile about legacy header file removals

jimklimov commented:
A couple of notable peculiarities:
* The build recipe (both in upstream and here) still refers to `BUILD_32` and so on, although the focus is more and more on delivering 64-bit only in some areas (e.g. kernel modules).
* Copy-pasted p5m files refer to NVidia Quattro drivers, I am not sure and did not check if this nomination is still valid ;)

jimklimov committed:
nvidia-304/340/384/387 : bump sample-manifest.p5m* about legacy header file removals

jimklimov committed:
nvidia-384/387 : bump Makefile and *.p5m* about 32-bit gfx_private driver removal

jimklimov committed:
nvidia-390/396 : introduce recipes

jimklimov committed:
nvidia-396 : deliver "egpu.html" documentation file

jimklimov committed:
nvidia metapackage : update sample-manifest.p5m copyright year

jimklimov committed:
nvidia-select : update comments; require that an OLD_DRV_ID is present when unloading

jimklimov committed:
nvidia-select : introduce ACTUAL_DRV_PATH and LOAD_DRV_SKIP_ABSENT=true support (for modules not delivered with every nvidia-NNN package variant)

kenmays commented:
Main driver versions to use as of today:

>     nvidia-304 - 304.137
>     nvidia-340 - 340.107
>     nvidia-440 - 440.31

This covers the main Nvidia graphics product releases from Y2004->Y2019. **The Nvidia PCI ID scanning starts from the most current video driver and _fallbacks_ to the older driver until found or will fallback to the _VESA_ driver if not found at all within the provided drivers.**

pyhalov commented on components/openindiana/nvidia-select/files/nvidia-select:
```diff
@@ -0,0 +1,415 @@
+# This should be executed as root (or sufficiently privileged user).
+#
+
+#
```
Please, move CDDL header before comment.

jimklimov replied:
ack, will do

pyhalov commented on components/openindiana/nvidia-select/files/nvidia-select:
```diff
@@ -0,0 +1,415 @@
+# http://www.illumos.org/license/CDDL.
+#
+# Copyright 2017, Jim Klimov
+# Portions Copyright 2003-2005 Sun Microsystems, Inc.  All rights reserved.
```
I'm not sure about 'Portions Copyright' string. Do we have precedents?
I'd just put Sun copyright on top, if this file has Sun heritage.

jimklimov replied:
Gotta remember now, not sure I can after so long :)

I'd guess this started by looking at `ogl-select` and at least modelling after the idea.

jimklimov replied:
Looking at current `ogl-select`, they are very dissimilar. So I must have started from something else back then :)

As for precedents, we have a few:

````
jim@jimoi:~/shared/oi-userland-tmp/components$ git grep -i 'portions copyright' | grep -v yyyy | wc -l
32
jim@jimoi:~/shared/oi-userland-tmp/components$ git grep -i 'portions copyright' | wc -l
1132
````

pyhalov commented on components/openindiana/nvidia-select/files/nvidia-select:
```diff
@@ -0,0 +1,415 @@
+    cmd="$cmd $drvname"
+    drvname="`echo $drvname | /usr/bin/sed 's;.*/;;g'`"
+
+    /usr/bin/grep "^$drvname[   ]" $BASEDIR/etc/name_to_major >  /dev/null 2>&1
```
What is it for?

jimklimov commented:
@kenmays : addition of nvidia-440 in progress locally, lots of different file lists to adapt the recipes iteratively :)

jimklimov committed:
Introduce nvidia-440

jimklimov committed:
nvidia-select script : move license to top of text

jimklimov commented:
@kenmays : added nvidia-440 to the wad. Not sure if I have anything to do in practice about your comment on PCI IDs : as far as I can see, the original driver packages assume only one copy/version present and active in the system, so either there are PCI IDs listed that this one supports, or not.

This PR (and `nvidia-select` in particular) should allow switching the symlinked names, and ID-DRV mappings, to the currently chosen active driver version -- if someone has several nvidia-NNN packages installed for experiments, upgrades... eventually as a default preinstalled footprint or Live env maybe?

kenmays commented:
@jimklimov - I concur with the default footprint for Live DVD Env as a test method as I still have GeForce Go 7600/Quadro FX 5500 for legacy driver testing. If things go right, I can install a Live DVD and it'll pick the right driver for these old graphics cards without any user/human interaction (but we can bulletproof this through a future startup screen selection/installation script - if needed). If everything goes right, anything from Quadro RTX 8000 -> GeForce 6000-series is properly detected and driver selected.

> **Nvidia driver coverage:** 
> Nvidia 440.x - covers GeForce RTX 20 series -> GeForce 400 series (primary)
> Nvidia 340.x - covers GeForce 300 series-> GeForce 8000 series (secondary)
> Nvidia 304.x - covers GeForce 7000 series -> GeForce 6000 series (tertiary)

I think the FreeBSD people are listening as they just updated their Nvidia driver revisions as well today:
See: https://www.freebsd.org/cgi/ports.cgi?query=nvidia&stype=name

The more, the merrier! ;>

jimklimov commented:
@kenmays : Any chance you could test this PR on hardware and/or propose the improvement implementations? :)

I might lack time (f)or imagination at the moment, but do not very well see how this current solution can help auto-detect the actual driver version to use... unless I made it two years ago already and forgot, Gallagher-style? :-)

I mean, what (IIRC) the `nvidia-select` script and service do is, essentially, ensure that the verbatim `nvidia` component of filenames (X11 and kernel driver) and directory names (a lot of these) points to one particular `nvidia-NNN`, and that this version's PCI IDs are recorded into system setup.

Thinking about it, if I get correctly what you could mean above, the script could probably look into the PCI ID mappings available in each installed `nvidia-NNN` package and into PCI IDs of the current hardware, (and track that it is in some auto mode vs. version preference set by user explicitly), and based on this pick a best-match video driver version. At least, this automagic makes sense and sounds neat, but I am not sure if it is in the code yet (and won't have time this week to look into it).

despair86 commented:
I've got a GTX 1650, so i practically live on bleeding-edge (and suffer bugs)

jimklimov commented:
Did anyone have a chance to test this on real HW that is supported by the drivers, so merging the PR can proceed?

I see that a new PR to piecemeal bump the one driver provided by the distro was posted recently by @mnowak , and fear that in a few days i'd have to solve new merge conflicts in this PR... and later again, and again, seems in vain ;)

kenmays commented:
@jimklimov - Yes, tested the Nvidia 440.44 driver w/RTX 2080 & GT 630 cards. Hope we can move forward.

mnowak commented:
> I see that a new PR to piecemeal bump the one driver provided by the distro was posted recently by @mnowak , and fear that in a few days i'd have to solve new merge conflicts in this PR... and later again, and again, seems in vain ;)

You mentioned the wrong guy pal :)

pyhalov commented on components/openindiana/nvidia-304/Makefile:
```diff
@@ -0,0 +1,202 @@
+install_mv_desktop_32:	install_reloc_32
+	mv -f "$(PROTO_DIR)/usr/share/applications/nvidia-settings.desktop" \
+	    "$(PROTO_DIR)/usr/share/applications/nvidia-settings-$(NVIDIA_DRIVER_MAJOR_VERSION).desktop" \
+	    && ln -s "nvidia-settings-$(NVIDIA_DRIVER_MAJOR_VERSION).desktop" \
```
I suppose this will not work as expected, we'll have several desktop files instead of one. Perhaps, simplify this by moving them and share/doc, share/icons to usr/share/NVIDIA-$(NVIDIA_DRIVER_MAJOR_VERSION)/some-subdirectory ?

pyhalov commented on components/openindiana/nvidia-select/files/nvidia-select:
```diff
@@ -0,0 +1,415 @@
+                ;;
+            esac
+    done
+    shift "`/usr/bin/expr $OPTIND - 1`"
```
/bin/sh can handle `$(($OPTIND - 1))

pyhalov commented on components/openindiana/nvidia-select/files/nvidia-select:
```diff
@@ -0,0 +1,415 @@
+    cmd="$cmd $drvname"
+    drvname="`echo $drvname | /usr/bin/sed 's;.*/;;g'`"
+
+    /usr/bin/grep "^$drvname[   ]" $BASEDIR/etc/name_to_major >  /dev/null 2>&1
```
No need in full path, as we set PATH in the begining

pyhalov commented on components/openindiana/nvidia-select/files/nvidia-select:
```diff
@@ -0,0 +1,415 @@
+        then
+            for i in $alias
+            do
+                /usr/bin/egrep "^$drvname[  ]+$i" $BASEDIR/etc/driver_aliases>/dev/null 2>&1
```
egrep -q ... 2>&1?

pyhalov replied:
or perhaps even avoid masking egrep errors

pyhalov commented on components/openindiana/nvidia-select/files/nvidia-select:
```diff
@@ -0,0 +1,415 @@
+}
+
+reload_drv() {
+    unload_drv "$@"
```
Can we really unload nvidia driver?

pyhalov commented on components/openindiana/nvidia-select/files/nvidia-select:
```diff
@@ -0,0 +1,415 @@
+    NVIDIA_VERSION_SMF="`svcprop -p options/NVIDIA_VERSION $SMF_FMRI`" \
+        && [ -n "$NVIDIA_VERSION_SMF" ] && NVIDIA_VERSION="$NVIDIA_VERSION_SMF"
+
+    if [ "$SMF_ACTION" = start ] \
```
What if we set options/NVIDIA_VERSION and reboot? 
Will we skip driver update?

pyhalov commented on components/openindiana/nvidia-select/files/nvidia-select:
```diff
@@ -0,0 +1,415 @@
+
+if [ -z "${EXISTING_MEDIATOR_DATA-}" ] ; then
+    echo "INFO: setting mediator $MEDIATOR_NAME to $MEDIATOR_IMPL / $NVIDIA_VERSION under $BASEDIR ..." >&2
+    $ALTROOT_FLAG /usr/bin/pkg set-mediator \
```
should we use -R option instead of chroot for pkg?

pyhalov commented:
Generally, this PR has one issue - overcomplicated installation logic for nvidia. 
Earlier we did just
```
$(INSTALL_32):  $(BUILD_32)
        [ -d $(PROTO_DIR)/kernel/drv/amd64 ] || mkdir -p $(PROTO_DIR)/kernel/drv/amd64
        for i in kernel/drv/amd64/nvidia kernel/drv/nvidia.conf; do \
            cp $(SOURCE_DIR)/NVDAgraphicsr/reloc/$$i $(PROTO_DIR)/$$i; done
        rm -fr $(PROTO_DIR)/usr
        cp -a $(SOURCE_DIR)/NVDAgraphics/reloc $(PROTO_DIR)/usr
        $(COMPONENT_POST_INSTALL_ACTION)
        $(TOUCH) $@
```
Now we have several steps, including moving files to separate dirs, moving individual files and so on. This makes updates to new major version not so straightforward. Can we somehow simplify these steps? Like move everything to /usr/share/nvidia-XXX and create necessary symlinks in several loops?

3eka commented:
I am running NVIDIA driver 440.31,REV=2019.10.27.02.19 (from nvidia site) on GK208B [GeForce GT 730] card, with no problem

kenmays commented:
Bump driver/graphics/nvidia to 440.44 in OI-IPS. Make it the new default.
Repackaged the legacy drivers in OI-IPS as: driver/graphics/nvidiaR304,  driver/graphics/nvidiaR340
Closes #5337.
The Nvidia 440.40 driver provides 10 years of legacy support back to GeForce 400-series. We can then simplify this process much easier by just focusing on the one driver update - instead of three!

Mno-hime commented:
I really don't have the skills to review nor understand, the less to maintain this change.

I think we should bite the bullet and plainly and transparently update to a sensible new default, be it version 390 (https://github.com/OpenIndiana/oi-userland/pull/5337), which I researched to be OK for almost everyone, or 440, the default in recent Linux distros.
